src/db.c

```diff
@@ -353,11 +417,80 @@ robj *dbRandomKey(redisDb *db) {
     }
 }
 
+/* Updates binary index tree (also known as Fenwick tree), increasing key count for a given slot.
+ * You can read more about this data structure here https://en.wikipedia.org/wiki/Fenwick_tree
+ * Time complexity is O(log(CLUSTER_SLOTS)). */
+void cumulativeKeyCountAdd(redisDb *db, int slot, long delta) {
+    if (!server.cluster_enabled) return;
+    int idx = slot + 1; /* Unlike slots, BIT is 1-based, so we need to add 1. */
+    while (idx <= CLUSTER_SLOTS) {
+        db->slot_size_index[idx] += delta;
+        idx += (idx & -idx);
+    }
+}
+
+/* Returns total (cumulative) number of keys up until given slot (inclusive).
+ * Time complexity is O(log(CLUSTER_SLOTS)). */
+unsigned long long cumulativeKeyCountRead(redisDb *db, int slot) {
+    if (!server.cluster_enabled) {
+        serverAssert(slot == 0);
+        return dbSize(db);
+    }
+    int idx = slot + 1;
+    unsigned long long sum = 0;
+    while (idx > 0) {
+        sum += db->slot_size_index[idx];
+        idx -= (idx & -idx);
+    }
+    return sum;
+}
+
+/* Returns fair random dictionary, probability of each dictionary being returned is proportional to the number of elements that it holds.
+ * Implementation uses binary search on top of binary index tree.
+ * Time complexity of this function is O(log^2(CLUSTER_SLOTS)). */
+dict *getFairRandomDict(redisDb *db) {
+    unsigned long target = dbSize(db) ? (randomULong() % dbSize(db)) + 1 : 0;
+    int slot = findSlotByKeyIndex(db, target);
+    return db->dict[slot];
+}
+
+/* Finds a slot containing target element in a key space ordered by slot id.
+ * Consider this example. Slots are represented by brackets and keys by dots:
+ *  #0   #1   #2     #3    #4
+ * [..][....][...][.......][.]
+ *                    ^
+ *                 target
+ *
+ * In this case slot #3 contains key that we are trying to find.
+ * */
+int findSlotByKeyIndex(redisDb *db, unsigned long target) {
+    if (!server.cluster_enabled || dbSize(db) == 0) return 0;
+    serverAssert(target <= dbSize(db));
+    int lo = 0, hi = CLUSTER_SLOTS - 1;
+    /* We use binary search to find a slot, we are allowed to do this, because we have a quick way to find a total number of keys
+     * up until certain slot, using binary index tree. */
+    while (lo <= hi) {
```

Comment:
It seems like we could utilize the binary lifting strategy instead of binary search for performing searches, which has the potential to optimize the complexity to log(n). [FYI](https://oi-wiki.org/ds/fenwick/#单点修改查询全局第-k-小).

Reply:
Even if we don't implement this, let's follow up. I'm not sure I followed the link, but I do intuitively believe that finding a slot by index should ~O(Log(N))